core/sr-io/without_std.rs

```diff
@@ -43,41 +51,214 @@ pub extern fn oom(_: ::core::alloc::Layout) -> ! {
 	static OOM_MSG: &str = "Runtime memory exhausted. Aborting";
 
 	unsafe {
-		ext_print_utf8(OOM_MSG.as_ptr(), OOM_MSG.len() as u32);
+		extern_functions_host_impl::ext_print_utf8(OOM_MSG.as_ptr(), OOM_MSG.len() as u32);
 		intrinsics::abort();
 	}
 }
 
+/// A function which implementation can be exchanged.
+///
+/// Internally this works by swapping function pointers.
+pub struct ExchangeableFunction<T>(T);
+
+impl<T> ExchangeableFunction<T> {
+	/// Replace the implementation with `new_impl`.
+	pub fn replace_implementation(self: &'static mut Self, new_impl: T)  -> RestoreImplementation<T> {
+		let old = mem::replace(&mut self.0, new_impl);
+
+		RestoreImplementation(self, Some(old))
+	}
+
+	/// Set the given implementation.
+	fn set_implementation(&mut self, new_impl: T) {
+		self.0 = new_impl;
+	}
+}
+
+impl<T> Deref for ExchangeableFunction<T> {
+	type Target = T;
+
+	fn deref(&self) -> &Self::Target {
+		&self.0
+	}
+}
+
+/// Restores a function implementation on drop.
+pub struct RestoreImplementation<T: 'static>(&'static mut ExchangeableFunction<T>, Option<T>);
+
+impl<T> Drop for RestoreImplementation<T> {
+	fn drop(&mut self) {
+		self.0.set_implementation(self.1.take().expect("Value is only taken on drop; qed"));
+	}
+}
+
+/// Declare extern functions
+macro_rules! extern_functions {
+	(
+		$(
+			$( #[$attr:meta] )*
+			fn $name:ident ( $( $arg:ident : $arg_ty:ty ),* ) $( -> $ret:ty )?;
+		)*
+	) => {
+		$(
+			$( #[$attr] )*
+			#[allow(non_upper_case_globals)]
+			pub static mut $name: ExchangeableFunction<unsafe fn ( $( $arg_ty ),* ) $( -> $ret )?> =
```

Comment:
what about threading?

Reply:
unsafe code in the runtime implementation terrifies me. Calling externs is one thing (this is where all unsafe to-date has been), but static mutable memory is another story.

We should expose a safe wrapper around this so when we do expose some kind of threading API we don't destroy everything.